src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.Core.cs

```diff
@@ -18,58 +18,71 @@
+        }
 
-        internal AnalyzerAssemblyLoader(ImmutableArray<IAnalyzerAssemblyResolver> externalResolvers)
-            : this(null, AnalyzerLoadOption.LoadFromDisk, externalResolvers)
+        internal AnalyzerAssemblyLoader(ImmutableArray<IAnalyzerPathResolver> pathResolvers)
+            : this(pathResolvers, assemblyResolvers: [DiskAnalyzerAssemblyResolver], compilerLoadContext: null)
         {
         }
 
-        internal AnalyzerAssemblyLoader(AssemblyLoadContext? compilerLoadContext, AnalyzerLoadOption loadOption, ImmutableArray<IAnalyzerAssemblyResolver> externalResolvers)
+        /// <summary>
+        /// Create a new <see cref="AnalyzerAssemblyLoader"/> with the given resolvers.
+        /// </summary>
+        /// <param name="compilerLoadContext">This is the <see cref="AssemblyLoadContext"/> where the compiler resides. This parameter
+        /// is primarily used for testing purposes but is also useful in hosted scenarios where the compiler may be loaded outside
+        /// the default context. When null this will be the <see cref="AssemblyLoadContext"/> the compiler currently resides
+        /// in </param>
+        /// <exception cref="ArgumentException"></exception>
+        internal AnalyzerAssemblyLoader(
+            ImmutableArray<IAnalyzerPathResolver> pathResolvers,
+            ImmutableArray<IAnalyzerAssemblyResolver> assemblyResolvers,
+            AssemblyLoadContext? compilerLoadContext)
         {
-            _loadOption = loadOption;
-            _compilerLoadContext = compilerLoadContext ?? AssemblyLoadContext.GetLoadContext(typeof(AnalyzerAssemblyLoader).GetTypeInfo().Assembly)!;
-            _externalResolvers = [.. externalResolvers, new CompilerAnalyzerAssemblyResolver(_compilerLoadContext)];
+            if (assemblyResolvers.Length == 0)
+            {
+                throw new ArgumentException("Cannot be empty", nameof(assemblyResolvers));
+            }
+
+            CompilerLoadContext = compilerLoadContext ?? AssemblyLoadContext.GetLoadContext(typeof(SyntaxTree).GetTypeInfo().Assembly)!;
+            CompilerAnalyzerAssemblyResolver = new CompilerResolver(CompilerLoadContext);
+            AnalyzerPathResolvers = pathResolvers;
+
+            // The CompilerAnalyzerAssemblyResolver must be first here as the host is _always_ given a chance
+            // to resolve the assembly before any other resolver. This is crucial to allow for items like
+            // unification of System.Collections.Immutable or other core assemblies for a host.
+            AnalyzerAssemblyResolvers = [CompilerAnalyzerAssemblyResolver, .. assemblyResolvers];
         }
 
         public bool IsHostAssembly(Assembly assembly)
         {
             CheckIfDisposed();
 
             var alc = AssemblyLoadContext.GetLoadContext(assembly);
-            return alc == _compilerLoadContext || alc == AssemblyLoadContext.Default;
+            return alc == CompilerLoadContext || alc == AssemblyLoadContext.Default;
```

Comment:
Not a question for this PR, but I'm surprised to see AssemblyLoadContext.Default being mentioned here since it's fairly ambiguous what that might be in more complicated scenarios. (I'm reminded of cases where MSBuild assumes it's something and that causes some pain too.)

Reply:
I tried making this an explicit v.s optional parameter to force the issue of "you must be specific about this" but the churn got a bit too big. And honestly, in all our production hosting now `Deafult` is the compiler context.